addons/l10n_pt_stock/models/stock_picking.py

```diff
@@ -0,0 +1,62 @@
+from odoo import models, api, fields, _
+from odoo.exceptions import UserError
+from odoo.tools import float_repr
+
+
+class Picking(models.Model):
+    _name = "stock.picking"
+    _inherit = ['stock.picking', 'hash.mixin']
+
+    invoice_no = fields.Char(string='InvoiceNo', compute='_compute_invoice_no')
+
+    @api.depends('name')
+    def _compute_invoice_no(self):
+        for picking in self:
+            name_split = picking.name.split('/')
+            picking.invoice_no = f'{name_split[0]} {name_split[1]}/{name_split[2]}'
+            pass
+
+    def _action_done(self):
+        super()._action_done()
+        if self.company_id.country_id.code == 'PT' and self.picking_type_id.code == 'outgoing':
+            for picking in self:
+                picking.inalterable_hash = picking._compute_hash(picking._get_previous_hash())
+
+    def _get_previous_hash(self):
+        self.ensure_one()
+        prev_picking = self.search([('state', '=', 'done'),
+                                    ('id', '!=', self.id),
+                                    ('picking_type_id.code', '=', 'outgoing'),
+                                    ('company_id', '=', self.company_id.id)],
+                                   limit=1,
+                                   order='date_done DESC')
+        if prev_picking and len(prev_picking) != 1:
+            raise UserError(
+               _('An error occured when computing the inalterability. Impossible to get the unique previous posted delivery note.'))
+        return prev_picking.inalterable_hash if prev_picking else ""
+
+    def _compute_hash(self, previous_hash):
+        self.ensure_one()
+        date_done = self.date_done.strftime('%Y-%m-%d')
+        system_entry_date = self.create_date.strftime("%Y-%m-%dT%H:%M:%S")
+        gross_total = float_repr(0, 2)
+        message = f"{date_done};{system_entry_date};{self.invoice_no};{gross_total};{previous_hash}"
+        return self.l10n_pt_compute_hash(message)
+
+    def l10n_pt_compute_hash(self, message):
+        """
+        This method's purpose is only to test that the hash is correctly
+        computed as we have multilple hash chains: one per move_type.
+        In each chain, we simply add one 1 to the previous hash value of that chain.
+        This method will be overriden in SaaS which will provide the real hash
+        """
+        self.ensure_one()
+        hash_string = int(message[message.rfind(';')+1:]) + 1 if message[-1].isdigit() else 0
+        return str(hash_string)
+
+    def write(self, vals):
+        if any(attr in vals for attr in ('state', 'date_done', 'name', 'picking_type_id', 'is_locked')):
+            for picking in self:
+                if picking.inalterable_hash:
+                    raise UserError(_("You cannot modify a validated delivery note."))
+        return super().write(vals)
```

Comment:
This should be done in the mixin (with the list of fields being `_get_fields_used_by_hash`)